de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateBL.java

```diff
@@ -1012,31 +1013,64 @@ else if (diffPercent.signum() < 0 && diffPercent.abs().compareTo(conditions.getM
 	public void extendContract(final @NonNull ContractExtendingRequest context)
 	{
 		Services.get(ITrxManager.class).run(ITrx.TRXNAME_ThreadInherited, localTrxName -> {
-			extendContract0(context, localTrxName);
 
-			final I_C_Flatrate_Term currentTerm = context.getContract();
-			currentTerm.setAD_PInstance_EndOfTerm_ID(context.getAD_PInstance_ID());
-			InterfaceWrapperHelper.save(currentTerm);
-
-			final I_C_Flatrate_Term nextTerm = currentTerm.getC_FlatrateTerm_Next();
-			final I_C_Flatrate_Conditions nextConditions = nextTerm.getC_Flatrate_Conditions();
-			final I_C_Flatrate_Transition nextTransition = nextConditions.getC_Flatrate_Transition();
-			if (nextTransition.isAutoExtension() && nextTransition.getC_Flatrate_Conditions_Next_ID() > 0)
+			ContractExtendingRequest contextUsed = context;
+			I_C_Flatrate_Transition nextTransition = null;
+			final List<I_C_Flatrate_Term> contracts = new ArrayList<>();
+			contracts.add(contextUsed.getContract());
+			do
 			{
-				final ContractExtendingRequest nextContext = ContractExtendingRequest.builder()
+				extendContract0(contextUsed, localTrxName);
+
+				final I_C_Flatrate_Term currentTerm = contextUsed.getContract();
+				currentTerm.setAD_PInstance_EndOfTerm_ID(contextUsed.getAD_PInstance_ID());
+				InterfaceWrapperHelper.save(currentTerm);
+
+				final I_C_Flatrate_Term nextTerm = currentTerm.getC_FlatrateTerm_Next();
+				final I_C_Flatrate_Conditions nextConditions = nextTerm.getC_Flatrate_Conditions();
+				Check.assumeNotNull(nextConditions, "C_Flatrate_Conditions shall not be null!");
+
+				nextTransition = nextConditions.getC_Flatrate_Transition();
+				Check.assumeNotNull(nextTransition, "C_Flatrate_Transition shall not be null!");
+
+				contextUsed = ContractExtendingRequest.builder()
 						.AD_PInstance_ID(context.getAD_PInstance_ID())
 						.contract(nextTerm)
```

Comment:
if you annotate ContractExtendingRequest with `@Builder(toBuilder=true), you should be able to do 
```java
contextUsed = contextUsed.toBuilder()
  .AD_PInstance_ID(context.getAD_PInstance_ID())
  .contract(nextTerm).build();
```
and avoid the properties that don't change

Reply:
Nice!
I did not know